### PR TITLE
Cover TemplateSamples with tests

### PR DESCRIPTION
**Pull request by YuliiaKovalova**
Related to https://github.com/dotnet/templating/issues/2522#event-7855091718

### CODE REVIEW

File: test/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests/TemplateEngineSamplesTest.cs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        [InlineData("15-computed-symbol", "sample15", null, "usage computed symbols")]
+        [InlineData("16-string-value-transform", "sample16", null, "usage of string value transformator")]
+        public async void TemplateEngineSamplesProjectTest(
+            string folderName,
+            string shortName,
+            string[] args,
+            string caseDescription)
+        {
+            _log.LogInformation($"Template with {caseDescription}");
+            string workingDir = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName().Replace(".", string.Empty));
```
Comment (JanKrivanek) [inline test/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests/TemplateEngineSamplesTest.cs:46]: tl;dr;: Remove this and it's usage

The testing FW inferes sensible temp working directory if it's not specified and then cleanup its content after test is run.
Custom specification is advisable if for some reasons you need the instantiation outputs to survive the test run (usually just investigation purposes).

File: test/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests/TemplateEngineSamplesTest.cs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+            TemplateVerifierOptions options = new TemplateVerifierOptions(templateName: shortName)
+            {
+                TemplatePath = templateLocation,
+                OutputDirectory = workingDir
+            }
+             .WithInstantiationThroughTemplateCreatorApi(GetTemplateArgs(args))
+             .WithCustomScrubbers(
+                ScrubbersDefinition.Empty
+                .AddScrubber(sb => sb.Replace(DateTime.Now.ToString("MM/dd/yyyy"), "**/**/****"))
+                .AddScrubber(sb => sb.Replace(DateTime.Now.ToString("MM/dd/yyyy"), "**/**/****")));
```
Comment (JanKrivanek) [inline test/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests/TemplateEngineSamplesTest.cs:60]: Copy paste error?

I'm not good in finding differences though :-)

File: test/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests/TemplateEngineSamplesTest.cs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+            VerificationEngine engine = new VerificationEngine(_log);
+            await engine.Execute(options);
```
Comment (JanKrivanek) [inline test/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests/TemplateEngineSamplesTest.cs:63]: nit: it's good practice to add '.ConfigureAwait(false)' to all `awaits` that do not need continuation on a same thread (usually needed only for GUI operations)

File: test/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests/TemplateEngineSamplesTest.cs [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+        public TemplateEngineSamplesTest(ITestOutputHelper log)
+        {
+            _log = new XunitLoggerProvider(log).CreateLogger("TestRun");
+        }
+
+        [Theory]
+        [InlineData("01-basic-template", "sample01", null, "no args")]
+        [InlineData("02-add-parameters", "sample02", new[] { "--copyrightName", "Test Copyright", "--Title", "Test Title" }, "text args")]
+        [InlineData("03-optional-page", "sample03", new[] { "--Title", "Test Title", "--EnableContactPage", "true" }, "optional content included")]
```
Comment (JanKrivanek) [inline test/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests/TemplateEngineSamplesTest.cs:24]: Does the test really accept the parameter? As calling through API requires params as defined in templateconfig model - so basically without the `--`.

I have feeling that the test works just coincidentaly (e.g. picking default value from config)

**Review by JanKrivanek — CHANGES_REQUESTED** [review]
Looks really great!
First application of the new testing FW - so no option to get proper inspiration (this code is going to be the new inspiration!) - good job!

I have few minor coments to make this even more shiny

File: test/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests.csproj [hunk 5]
```diff
@@ -22,4 +22,8 @@
     <Compile Include="..\Shared\**\*.cs" Link="%(RecursiveDir)%(Filename)%(Extension)" />
   </ItemGroup>
 
+  <ItemGroup>
+    <None Remove="Snapshots\TemplateEngineSamplesProjectTest.sample05._.verified\sample05\README.md" />
+  </ItemGroup>
+
```
Comment (vlada-shubina) [inline test/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests.csproj:28]:
```suggestion
```

File: test/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests/TemplateEngineSamplesTest.cs [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        }
+
+        [Theory]
+        [InlineData("01-basic-template", "sample01", null, "no args")]
+        [InlineData("02-add-parameters", "sample02", new[] { "copyrightName", "Test Copyright", "title", "Test Title" }, "text args")]
+        [InlineData("03-optional-page", "sample03", new[] { "enableContactPage", "true" }, "optional content included")]
+        [InlineData("03-optional-page", "sample03", null, "optional content excluded")]
+        [InlineData("04-parameter-from-list", "sample04", new[] { "BackgroundColor", "dimgray" }, "the parameter from the list of options")]
+        [InlineData("05-multi-project", "sample05", new[] { "--includetest", "true" }, "the optional test project included")]
+        [InlineData("05-multi-project", "sample05", new[] { "--includetest", "false" }, "the optional test project excluded")]
```
Comment (vlada-shubina) [inline test/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests/TemplateEngineSamplesTest.cs:28]:
```suggestion
        [InlineData("05-multi-project", "sample05", new[] { "includetest", "true" }, "the optional test project included")]
        [InlineData("05-multi-project", "sample05", new[] { "includetest", "false" }, "the optional test project excluded")]
```

when testing through API, the parameter names should match the name in `template.json`, i.e. `--` is not needed.

File: test/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests/TemplateEngineSamplesTest.cs [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        [InlineData("06-console-csharp-fsharp", "sample06", new[] { "--language", "F#" }, "the F# lang parameter that creates a corresponded project")]
+        [InlineData("06-console-csharp-fsharp", "sample06", new[] { "--language", "C#" }, "the C# lang parameter that creates a corresponded project")]
```
Comment (vlada-shubina) [inline test/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests/TemplateEngineSamplesTest.cs:30]: Language parameter only works in CLI, so passing it does nothing.  @JanKrivanek am I right here? is there a way to diff the language in API tests?

File: test/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests/TemplateEngineSamplesTest.cs [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+        [InlineData("07-param-with-custom-short-name", "sample07", new[] { "preferNameDirectory", "true" }, "custom name directory")]
```
Comment (vlada-shubina) [inline test/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests/TemplateEngineSamplesTest.cs:31]: `preferNameDirectory` is not a parameter, should it be something else?

File: test/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests/TemplateEngineSamplesTest.cs [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+        [InlineData("08-restore-on-create", "sample08", null, "restore on create")]
```
Comment (vlada-shubina) [inline test/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests/TemplateEngineSamplesTest.cs:32]: not actionable: restore won't happen in API test. Restore post action is defined in SDK. For this test, it will be also good to execute it via `dotnet new`, same as sample 5 and sample 7

File: test/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests/Snapshots/TemplateEngineSamplesProjectTest.sample02._.verified/sample02/Program.cs [hunk 10]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
```
Comment (vlada-shubina) [inline test/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests/Snapshots/TemplateEngineSamplesProjectTest.sample02._.verified/sample02/Program.cs:1]: these files have weird naming `._.` should we just define custom name here?

File: test/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests/TemplateEngineSamplesTest.cs [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+            string caseDescription)
+        {
+            _log.LogInformation($"Template with {caseDescription}");
+
+            //get the template location
+            string templateLocation = Path.Combine(GetSamplesTemplateLocation(), folderName);
+
+            TemplateVerifierOptions options = new TemplateVerifierOptions(templateName: shortName)
+            {
+                TemplatePath = templateLocation
```
Comment (vlada-shubina) [inline test/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests/TemplateEngineSamplesTest.cs:52]:
```suggestion
                TemplatePath = templateLocation,
                DoNotPrependCallerMethodNameToScenarioName = true,
                ScenarioName = folderName
                
```

for cleaner naming. If `folderName` is not good, please add another parameter to data set: `scenarioName` and use it.

File: test/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests/TemplateEngineSamplesTest.cs (hunk 7 above)
Comment (JanKrivanek) [reply test/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests/TemplateEngineSamplesTest.cs:30]: Yeah - unfortunately this is CLI functionality - so it would need to be replicated in the testing aPI extension - however it would be at least controversial if not boldly wrong. As we'd then test some functionality that's now in sdk, but we'd actualy test it's replicated version that's part of testing framework.

tl;dr; scenarios that need advanced CLI functionality should not be tested within templating repo, but rather sdk repo. Here (in templating), we need to explictly call each template separately

Comment (YuliiaKovalova) [reply test/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests/TemplateEngineSamplesTest.cs:30]: I remove sample06 from here.
It will be covered in sdk

File: test/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests/TemplateEngineSamplesTest.cs (hunk 8 above)
Comment (YuliiaKovalova) [reply test/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests/TemplateEngineSamplesTest.cs:31]: it is supposed to use values from dotnetcli.host.json
I remove this extra param.

File: test/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests/TemplateEngineSamplesTest.cs (hunk 7 above)
Comment (vlada-shubina) [reply test/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests/TemplateEngineSamplesTest.cs:30]: @JanKrivanek i think it would be good addition to API verifier though.
Use case: the package contains the templates for same language:
|Identity|short-name|language|
|---|---|---|
|MyTemplate.FSharp|mytemplate| F#|
|MyTemplate.CSharp|mytemplate| C#|

There is no way to execute specific one from that one. Probably extending [here](https://github.com/dotnet/templating/blob/bb19dc09e1c32c1fd75f3539dc75590c5a08e028/tools/Microsoft.TemplateEngine.Authoring.TemplateApiVerifier/TemplateVerifierOptionsExtensions.cs#L98) to allow passing custom filter may work.

Comment (vlada-shubina) [reply test/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests/TemplateEngineSamplesTest.cs:30]: Added this to https://github.com/dotnet/templating/issues/5705

File: test/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests/Snapshots/TemplateEngineSamplesProjectTest.sample02._.verified/sample02/Program.cs (hunk 10 above)
Comment (YuliiaKovalova) [reply test/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests/Snapshots/TemplateEngineSamplesProjectTest.sample02._.verified/sample02/Program.cs]: I have added custom naming for this test case
please take a look
@vlada-shubina , @JanKrivanek

Comment (JanKrivanek) [reply test/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests/Snapshots/TemplateEngineSamplesProjectTest.sample02._.verified/sample02/Program.cs]: Unfortnuately currently either encoded params, or underscore or custom namers are needed - as Verify requires nonempty MethodName and TypeName and those are then used to name the snapshot. So based on various naming options within TemplateVerifierOptions or in a case that there are no template params, we anyway sometimes need to end up making up some names.
I personaly do not feel the `_` in name should be problematic. It at least explicitly denotes tha fact that no params were specified for instantiation.

tl;dr; Specifying ScenarioName is acceptable workaround.

**Comment by YuliiaKovalova** [comment]
@vlada-shubina , @JanKrivanek 

I have addressed all your comments. Please take a look

File: test/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests/TemplateEngineSamplesTest.cs [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+        {
+            _log = new XunitLoggerProvider(log).CreateLogger("TestRun");
+        }
+
+        [Theory]
+        [InlineData("01-basic-template", "sample01", null, "no args")]
+        [InlineData("02-add-parameters", "sample02", new[] { "copyrightName", "Test Copyright", "title", "Test Title" }, "text args")]
+        [InlineData("03-optional-page", "sample03", new[] { "enableContactPage", "true" }, "optional content included")]
+        [InlineData("03-optional-page", "sample03", null, "optional content excluded")]
+        [InlineData("04-parameter-from-list", "sample04", new[] { "BackgroundColor", "dimgray" }, "the parameter from the list of options")]
```
Comment (vlada-shubina) [inline test/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests/TemplateEngineSamplesTest.cs:27]:
```suggestion
        [InlineData("04-parameter-from-list", "sample04", new[] { "BackgroundColor", "dimgray" }, "the choice parameter")]
```

File: test/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests/TemplateEngineSamplesTest.cs [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+        [InlineData("05-multi-project", "sample05", new[] { "includetest", "true" }, "the optional test project included")]
+        [InlineData("05-multi-project", "sample05", new[] { "includetest", "false" }, "the optional test project excluded")]
+        [InlineData("07-param-with-custom-short-name", "sample07", null, "customised name directory")]
```
Comment (vlada-shubina) [inline test/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests/TemplateEngineSamplesTest.cs:30]:
```suggestion
        [InlineData("07-param-with-custom-short-name", "sample07", null, "customised parameter name")]
```

File: test/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests/TemplateEngineSamplesTest.cs [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+        [InlineData("08-restore-on-create", "sample08", null, "restore on create")]
+        [InlineData("09-replace-onlyif-after", "sample09", new[] { "backgroundColor", "grey" }, "replacing with onlyif condition")]
+        [InlineData("10-symbol-from-date", "sample10", null, "usage of date generator")]
+        [InlineData("11-change-string-casing", "sample11", null, "usage of casing generator")]
+        [InlineData("13-constant-value", "sample13", null, "replacing of constant value")]
+        [InlineData("15-computed-symbol", "sample15", null, "usage computed symbols")]
+        [InlineData("16-string-value-transform", "sample16", null, "usage of string value transformator")]
```
Comment (vlada-shubina) [inline test/Microsoft.TemplateEngine.Authoring.TemplateVerifier.IntegrationTests/TemplateEngineSamplesTest.cs:37]:
```suggestion
        [InlineData("16-string-value-transform", "sample16", null, "usage of derived parameter")]
```

**Review by vlada-shubina — APPROVED** [review]
Looks good

**Review by JanKrivanek — APPROVED** [review]
Perfect!
